src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,189 @@
+                head = newNode;
+            }
+            size++;
+        } else if (index == size) {
+            newNode.prev = tail;
+            tail.next = newNode;
+            tail = newNode;
+            size++;
+        } else {
+            Node<T> current = head;
+            for (int i = 0; i < index - 1; i++) {
+                current = current.next;
+            }
+            current.next.prev = newNode;
+            newNode.next = current.next;
+            current.next = newNode;
+            newNode.prev = current;
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        Node<T> newNode;
+        if (list != null) {
+            for (T el : list) {
+                newNode = new Node<>(null, el, null);
+                if (head == null) {
+                    head = tail = newNode;
+                } else {
+                    tail.next = newNode;
+                    tail = newNode;
+                }
+                size++;
+            }
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (head == null || index >= size) {
+            throw new IndexOutOfBoundsException("Incorrect index " + index);
+        }
+        Node<T> current = head;
+        for (int i = 0; i < index; i++) {
+            current = current.next;
+        }
+        return current.value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Incorrect index " + index);
+        }
+        Node<T> current = head;
+        for (int i = 0; i < index; i++) {
+            current = current.next;
+        }
+        T tmp = current.value;
+        current.value = value;
+        return tmp;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Incorrect index " + index);
+        }
+        T tmp;
+        if (index == 0) {
+            tmp = head.value;
+            if (size == 1) {
+                head = null;
+                tail = null;
+            } else {
+                head = head.next;
+                head.prev = null;
+            }
+        } else if (index == size - 1) {
+            tmp = tail.value;
+            tail = tail.prev;
+            if (tail == null) {
+                head = null;
+            } else {
+                tail.next = null;
+            }
+        } else {
+            Node<T> current = head;
+            for (int i = 0; i < index; i++) {
+                current = current.next;
+            }
+            tmp = current.value;
+            current.prev.next = current.next;
+            current.next.prev = current.prev;
+        }
+        size--;
+        return tmp;
     }
 
     @Override
     public boolean remove(T object) {
+        if (head == null) {
```

Comment:
Don't forget that you can have `null` as a method parameter in the `remove(T object)` method. Make sure to handle the case where `object` is `null` correctly.